docs/source/developer/rewrites.rst

```diff
@@ -0,0 +1,394 @@
+   for Numba rewrites.  Developers should define rewrites as
+   subclasses of this base type, overloading the
+   :func:`~Rewrite.match` and :func:`~Rewrite.apply` methods.
+
+   .. attribute:: pipeline
+
+       The pipeline attribute contains the
+       :class:`numba.compiler.Pipeline` instance that is currently
+       compiling the function under consideration for rewriting.
+
+   .. method:: __init__(self, pipeline, *args, **kws)
+
+       The base constructor for rewrites simply stashes its arguments
+       into attributes of the same name.  Unless being used in
+       debugging or testing, rewrites should only be constructed by
+       the :class:`RewriteRegistry` in the
+       :func:`RewriteRegistry.apply` method, and the construction
+       interface should remain stable (though the pipeline will
+       commonly contain just about everything there is to know).
+
+   .. method:: match(self, block, typemap, callmap)
+
+      The :func:`~Rewrite.match` method takes three arguments other
+      than *self*:
+
+      * *block*: This is an instance of :class:`numba.ir.Block`.  The
+        matching method should iterate over the instructions contained
+        in the :attr:`numba.ir.Block.body` member.
+
+      * *typemap*: This is a Python :class:`dict` instance mapping
+        from symbol names in the IR, represented as strings, to Numba
+        types.
+
+      * *callmap*: This is another :class:`dict` instance mapping from
+        calls, represented as :class:`numba.ir.Expr` instances, to
+        their corresponding call site type signatures, represented as
+        a :class:`numba.typing.templates.Signature` instance.
+
+      The :func:`~Rewrite.match` method should return a :class:`bool`
+      result.  A :obj:`True` result should indicate that one or more
+      matches were found, and the :func:`~Rewrite.apply` method will
+      return a new replacement :class:`numba.ir.Block` instance.  A
+      :obj:`False` result should indicate that no matches were found, and
+      subsequent calls to :func:`~Rewrite.apply` will return undefined
+      or invalid results.
+
+   .. method:: apply(self)
+
+      The :func:`~Rewrite.apply` method should only be invoked
+      following a successful call to :func:`~Rewrite.match`.  This
+      method takes no additional parameters other than *self*, and
+      should return a replacement :class:`numba.ir.Block` instance.
+
+      As mentioned above, the behavior of calling
+      :func:`~Rewrite.apply` is undefined unless
+      :func:`~Rewrite.match` has already been called and returned
+      :obj:`True`.
+
+
+Subclassing :class:`Rewrite`
+----------------------------
+
+Before going into the expectations for the overloaded methods any
+:class:`Rewrite` subclass must have, let's step back a minute to
+review what is taking place here.  By providing an extensible
+compiler, Numba opens itself to user-defined code generators which may
```

Comment:
It is something we want to officially support, but it might be a little premature to advertise at the moment.  Part of the goal of the 0.20 release will be to do some refactoring and interface cleanup so that we can promise an extensible compiler interface.